### PR TITLE
Print warning to parallelize tests when applicable

### PR DESCRIPTION
**Pull request by plypaul**
### Description

This PR adds a warning if a test session could benefit from changing the parallelism.

<!--- 
  Before requesting review, please make sure you have:
  1. read [the contributing guide](https://github.com/dbt-labs/metricflow/blob/main/CONTRIBUTING.md),
  2. signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
  3. run `changie new` to [create a changelog entry](https://github.com/dbt-labs/metricflow/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: metricflow/sql_clients/sql_utils.py [hunk 1]
```diff
@@ -65,12 +65,17 @@ def make_df(  # type: ignore [misc]
     )
 
 
-def make_sql_client(url: str, password: str) -> AsyncSqlClient:
-    """Build SQL client based on env configs. Used only in tests."""
+def dialect_from_url(url: str) -> SqlDialect:
```
Comment (tlento) [inline metricflow/sql_clients/sql_utils.py:68]: Was not expecting this here, it'd be good to put the need for this change in the commit message.

File: metricflow/test/fixtures/setup_fixtures.py [hunk 2]
```diff
@@ -76,7 +80,7 @@ class MetricFlowTestEnvironmentVariables:
 def mf_test_session_state(  # noqa: D
     request: FixtureRequest,
     source_table_snapshot_repository: SqlTableSnapshotRepository,
-) -> MetricFlowTestSessionState:
+) -> Generator[MetricFlowTestSessionState, None, None]:
```
Comment (tlento) [inline metricflow/test/fixtures/setup_fixtures.py:83]: I'm very confused as to why this change "just works" - seems to me every call to this fixture should break since you were getting a MetricFlowTestSessionState and now you can't access it without an iterator call.

Is there some Python generator magic I'm missing here?  Or is this pytest fixture + generator magic?

File: metricflow/test/fixtures/setup_fixtures.py [hunk 3]
```diff
@@ -142,6 +146,35 @@ def mf_test_session_state(  # noqa: D
...
+    # 10/30 as the thresholds, but not much thought has been put into it. The following could also be a single
+    # expression, but the nested if structure seemed easier to read.
+    could_benefit_from_parallel_tests = False
+    if num_workers == 1:
+        if dialect is SqlDialect.DUCKDB or dialect is SqlDialect.POSTGRESQL:
+            if num_items > 30:
+                could_benefit_from_parallel_tests = True
+        elif num_items > 10:
+            could_benefit_from_parallel_tests = True
+
```
Comment (tlento) [inline metricflow/test/fixtures/setup_fixtures.py:169]: Huh, the threshold layering is a little clunky. Maybe we can do early returns to make those more obvious, something like the below?

Also I think notes about the arbitrary thresholds can go in a docstring instead of inline, that gets picked up by automation a bit better.

```suggestion
    if num_workers > 1 or num_items < 10:
        # Don't warn if we're already using parallelization or test suite is small
        return
    if (dialect is SqlDialect.DUCKDB or dialect is SqlDialect.POSTGRESQL) and num_items < 30:
        # parallelization threshold for DuckDB and Postgres is higher because they are fast
        return
        
```

**Review by tlento — APPROVED** [review]
Seems reasonable, left a readability suggestion for you.

I am baffled as to why tests relying on that session state fixture continue to pass with the switch from return to yield.

File: metricflow/test/fixtures/setup_fixtures.py (hunk 3 above)
Comment (plypaul) [reply metricflow/test/fixtures/setup_fixtures.py]: Updated.

File: metricflow/sql_clients/sql_utils.py (hunk 1 above)
Comment (plypaul) [reply metricflow/sql_clients/sql_utils.py:68]: Updated.
